File: packages/codex-session-relay/src/codex_session_relay/criteria.py
"""The canonical criteria a delivery is judged against, and the set a review was bound to.

Held here rather than in the verdict record because the five contract schemas are frozen with
additionalProperties false, so contract v1 has no slot for a criteria set. This is a relay-owned
record, like the revision request.

Two modes, named rather than implied. A MANAGED assignment is one somebody registered criteria
for, or declared managed: it cannot be completed as verified against nothing, because "verified
against nothing" is the claim this workflow exists to prevent. A LEGACY relationship is the
generic relay with no criteria contract, and it keeps the delivered behaviour exactly. The mode
is stored rather than inferred from whether a set happens to exist, because an absent set on a
managed assignment is precisely the case that must refuse.
"""

import hashlib
import json

from . import restoration
from .errors import AckRefused, RefusalReason

DISPOSITIONS = ("verified", "needs_changes", "unverified")
MANAGED = "managed"
LEGACY = "legacy"
MODES = (MANAGED, LEGACY)

COVERED = "covered"
LEGACY_UNREGISTERED = "legacy_unregistered"


def set_digest(entries) -> str:
    """sha256 over a canonical JSON serialisation, not over joined text.

    Delimiter-joined id|title|required cannot be made collision-free by escaping rules nobody
    will remember: a title containing the delimiter silently becomes a different set with the
    same digest. JSON quoting already escapes every delimiter, so two different sets cannot
    serialise identically.
    """
    payload = json.dumps(
        [
            {"id": e["id"], "title": e["title"], "required": bool(e.get("required", True))}
            for e in sorted(entries, key=lambda e: e["id"])
        ],
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def normalise_findings(criteria=None, findings=None) -> list:
    """One shape for what the CLI and the API both call findings.

    criteria is the delivered core's simple form, [{"id", "verdict"}]. findings adds a note.
    Both end up here, and a disposition outside the frozen enum is refused before anything is
    written, because a record that would fail conformance must never reach the store.

    A finding may also declare that it carries the correction's restoration block. That is the
    one place the block can travel, so which finding holds it has to survive this
    normalisation rather than being dropped with every other unrecognised key.
    """
    merged = []
    for source in (criteria or [], findings or []):
        for item in source:
            if not isinstance(item, dict):
                raise AckRefused(
                    RefusalReason.DISPOSITION_CONFLICT, "each finding is an object"
                )
            identifier = str(item.get("id") or "").strip()
            if not identifier:
                raise AckRefused(
                    RefusalReason.DISPOSITION_CONFLICT, "each finding names a criterion id"
                )
            disposition = item.get("verdict") or "verified"
            if disposition not in DISPOSITIONS:
                raise AckRefused(
                    RefusalReason.DISPOSITION_CONFLICT,
                    f"{disposition!r} is not one of {DISPOSITIONS}; the contract's criteria "
                    "enum is frozen and a finding outside it cannot be recorded",
                )
            entry = {"id": identifier, "verdict": disposition}
            note = str(item.get("note") or "").strip()
            if note:
                entry["note"] = note
            flag = item.get(restoration.FIELD)
            if flag is not None and not isinstance(flag, bool):
                raise AckRefused(
                    RefusalReason.DISPOSITION_CONFLICT,
                    f"a finding declares its restoration block with true or false, not "
                    f"{type(flag).__name__}",
                )
            previous = next((e for e in merged if e["id"] == identifier), None)
            declared_before = bool(previous and previous.get(restoration.FIELD))
            if flag is False and declared_before:
                # Last-entry-wins is right for a disposition and a note and wrong for this.
                # The two inputs are merged by id, so a caller that declared the block in
                # criteria and then described it in findings would have the declaration
                # cancelled by the entry that was only supposed to add the note. A caller
                # that means to cancel it cannot be told apart from one that forgot, so the
                # contradiction is refused rather than guessed either way.
                raise AckRefused(
                    RefusalReason.DISPOSITION_CONFLICT,
                    f"{identifier!r} both declares and disclaims the restoration block; one "
                    "correction carries one block and says so once",
                )
            if flag or declared_before:
                entry[restoration.FIELD] = True
            merged = [e for e in merged if e["id"] != identifier] + [entry]
    carriers = [e["id"] for e in merged if e.get(restoration.FIELD)]
    if len(carriers) > 1:
        raise AckRefused(
            RefusalReason.DISPOSITION_CONFLICT,
            f"{carriers} each declare the restoration block. One correction carries one "
            "block, and two candidates is a block nobody can locate",
        )
    # After the merge, so a note supplied through the other input still counts. The block
    # travels in the note and nowhere else, so a declaration with no note marks an empty
    # carrier: coverage is satisfied by some other actionable finding, the renderer labels the
    # empty one, and the relay reports it carried. Every check downstream tests the boolean,
    # so this is the only place that can tell an intention from a delivery.
    empty = [e["id"] for e in merged if e.get(restoration.FIELD) and not e.get("note")]
    if empty:
        raise AckRefused(
            RefusalReason.DISPOSITION_CONFLICT,
            f"{empty[0]!r} declares the restoration block and carries no note. The block is "
            "the note; a declaration without one names a carrier with nothing in it",
        )
    return merged


class CriteriaService:
    def __init__(self, store, clock):
        self.store = store
        self.clock = clock

    # ------------------------------------------------------------- registration

    def register(self, relationship_id, entries, *, source_ref=None) -> dict:
        normalised, seen = [], set()
        for entry in entries:
            identifier = str(entry.get("id") or "").strip()
            title = str(entry.get("title") or "").strip()
            if not identifier or not title:
                raise AckRefused(
                    RefusalReason.CRITERIA_UNREGISTERED,
                    "each criterion needs a non-empty id and title",
                )
            if identifier in seen:
                raise AckRefused(
                    RefusalReason.CRITERIA_UNREGISTERED,
                    f"duplicate criterion id {identifier!r}",
                )
            seen.add(identifier)
            normalised.append(
                {"id": identifier, "title": title, "required": bool(entry.get("required", True))}
            )
        if not normalised:
            raise AckRefused(
                RefusalReason.CRITERIA_UNREGISTERED,
                "a criteria set needs at least one criterion",
            )
        digest = set_digest(normalised)
        now = self.clock.iso()
        with self.store.transaction() as db:
            db.execute(
                "DELETE FROM canonical_criteria WHERE relationship_id = ?", (relationship_id,)
            )
            for entry in normalised:
                db.execute(
                    "INSERT INTO canonical_criteria (relationship_id, criterion_id, title,"
                    " required, source_ref, set_digest, recorded_at) VALUES (?,?,?,?,?,?,?)",
                    (
                        relationship_id, entry["id"], entry["title"], int(entry["required"]),
                        source_ref, digest, now,
                    ),
                )
            self._write_mode(db, relationship_id, MANAGED, now)
            self.store.journal(
                "criteria_registered", relationship_id,
                {"setDigest": digest, "count": len(normalised)}, at=now,
            )
        return {
            "relationshipId": relationship_id, "setDigest": digest, "sourceRef": source_ref,
            "criteria": normalised, "mode": MANAGED,
        }

    def get(self, relationship_id):
        rows = self.store.all(
            "SELECT * FROM canonical_criteria WHERE relationship_id = ? ORDER BY criterion_id",
            (relationship_id,),
        )
        if not rows:
            return None
        return {
            "relationshipId": relationship_id,
            "setDigest": rows[0]["set_digest"],
            "sourceRef": rows[0]["source_ref"],
            "criteria": [
                {"id": r["criterion_id"], "title": r["title"], "required": bool(r["required"])}
                for r in rows
            ],
        }

    def mode(self, relationship_id) -> str:
        row = self.store.one(
            "SELECT mode FROM verification_mode WHERE relationship_id = ?", (relationship_id,)
        )
        return row["mode"] if row else LEGACY

    def set_mode(self, relationship_id, mode) -> dict:
        if mode not in MODES:
            raise AckRefused(
                RefusalReason.DISPOSITION_CONFLICT, f"unknown verification mode {mode!r}"
            )
        now = self.clock.iso()
        with self.store.transaction() as db:
            self._write_mode(db, relationship_id, mode, now)
        return {"relationshipId": relationship_id, "mode": mode}

    def _write_mode(self, db, relationship_id, mode, now) -> None:
        db.execute(
            "INSERT INTO verification_mode (relationship_id, mode, recorded_at) VALUES (?,?,?)"
            " ON CONFLICT(relationship_id) DO UPDATE SET mode = excluded.mode,"
            " recorded_at = excluded.recorded_at",
            (relationship_id, mode, now),
        )

    # ------------------------------------------------------------------ binding

    def bind_review(self, db, relationship_id, event_id):
        """Record the digest of the set as it stands AT REVIEW START.

        This is what makes editing a criterion's text while keeping its id invalidate the
        review instead of silently passing it: findings made against one wording cannot be
        used to certify a different wording.
        """
        current = self.get(relationship_id)
        digest = current["setDigest"] if current else None
        db.execute(
            "INSERT OR IGNORE INTO claim_context (event_id, set_digest, bound_at) VALUES (?,?,?)",
            (event_id, digest, self.clock.iso()),
        )
        return digest

    def bound_digest(self, event_id):
        row = self.store.one(
            "SELECT set_digest FROM claim_context WHERE event_id = ?", (event_id,)
        )
        return row["set_digest"] if row else None

    # ----------------------------------------------------------------- coverage

    def coverage(self, relationship_id, event_id, verdict, findings, *, reason=None,
                 expected_digest=None) -> dict:
        registered = self.get(relationship_id)
        mode = self.mode(relationship_id)
        findings = findings or []

        if registered is None:
            if mode == MANAGED:
                raise AckRefused(
                    RefusalReason.CRITERIA_UNREGISTERED,
                    f"{relationship_id!r} is a managed assignment with no canonical criteria; "
                    "a managed assignment cannot be completed against nothing",
                )
            return {
                "coverage": LEGACY_UNREGISTERED, "setDigest": None, "boundDigest": None,
                "findings": findings,
            }

        digest = registered["setDigest"]
        bound = self.bound_digest(event_id)
        if (mode == MANAGED and bound is None and expected_digest is None
                and verdict in ("verified", "needs_changes")):
            # Without a bound review or a stated digest there is nothing to compare the
            # current set against, so skipping the claim would quietly skip the currency
            # protection entirely. Claiming the review is what pins the set it was made
            # against; stating the digest is the explicit alternative for a caller that has
            # one. Legacy relationships are unaffected and keep the delivered behaviour.
            raise AckRefused(
                RefusalReason.REVIEW_NOT_BOUND,
                f"this managed review is not bound to a criteria set: claim the event first, "
                f"or pass the reviewed digest explicitly. Current set is {digest}",
            )
        if bound is not None and bound != digest:
            raise AckRefused(
                RefusalReason.CRITERIA_SET_CHANGED,
                f"the criteria set changed after this review was claimed: bound {bound}, "
                f"current {digest}. Findings made against the previous wording cannot certify "
                "the current one; claim the review again",
            )
        if expected_digest is not None and expected_digest != digest:
            raise AckRefused(
                RefusalReason.CRITERIA_SET_CHANGED,
                f"expected criteria set {expected_digest}, but the current set is {digest}",
            )

        known = {c["id"] for c in registered["criteria"]}
        required = {c["id"] for c in registered["criteria"] if c["required"]}
        by_id = {}
        for finding in findings:
            if finding["id"] not in known:
                raise AckRefused(
                    RefusalReason.UNKNOWN_CRITERION,
                    f"{finding['id']!r} is not in this assignment's canonical criteria",
                )
            by_id[finding["id"]] = finding

        if verdict == "verified":
            passed = {i for i, f in by_id.items() if f["verdict"] == "verified"}
            missing = sorted(required - passed)
            if missing:
                raise AckRefused(
                    RefusalReason.CRITERIA_NOT_COVERED,
                    f"these required criteria are not recorded as verified: {missing}",
                )
        elif verdict == "needs_changes":
            actionable = [
                f for f in by_id.values()
                if f["verdict"] == "needs_changes" and f.get("note")
            ]
            if not actionable:
                raise AckRefused(
                    RefusalReason.FINDINGS_REQUIRED,
                    "a needs_changes verdict needs at least one criterion marked "
                    "needs_changes with a note; a correction with no findings is one nobody "
                    "can act on",
                )
        elif verdict == "unverified":
            stated = [
                f for f in by_id.values() if f["verdict"] == "unverified" and f.get("note")
            ]
            if not stated:
                raise AckRefused(
                    RefusalReason.FINDINGS_REQUIRED,
                    "an unverified verdict names at least one criterion it could not verify, "
                    "with the reason",
                )
        elif verdict == "aborted":
            if not str(reason or "").strip():
                raise AckRefused(
                    RefusalReason.FINDINGS_REQUIRED, "an aborted verdict states why"
                )

        return {
            "coverage": COVERED, "setDigest": digest, "boundDigest": bound, "findings": findings,
        }

File: internal/relay/delivery/criteria.go
package delivery

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"fmt"
	"slices"
	"sort"
	"strings"

	"github.com/thisisjun786/codex-relay-workflow/internal/relay/store"
)

// Criteria refusals and modes (criteria.py, errors.RefusalReason).
const (
	CriteriaUnregistered = "criteria_unregistered"
	CriteriaNotCovered   = "criteria_not_covered"
	CriteriaSetChanged   = "criteria_set_changed"
	UnknownCriterion     = "unknown_criterion"
	FindingsRequired     = "findings_required"
	DispositionConflict  = "disposition_conflict"
	ReviewNotBound       = "review_not_bound"
	Managed              = "managed"
	Legacy               = "legacy"
	Covered              = "covered"
	LegacyUnregistered   = "legacy_unregistered"
)

var dispositions = []string{"verified", "needs_changes", "unverified"}

// Criterion is one canonical criterion.
type Criterion struct {
	ID, Title string
	Required  bool
}

func (c Criterion) obj() Obj {
	return Obj{{Key: "id", Value: c.ID}, {Key: "title", Value: c.Title}, {Key: "required", Value: c.Required}}
}

func criteriaObjs(cs []Criterion) []any {
	out := make([]any, len(cs))
	for i, c := range cs {
		out[i] = c.obj()
	}
	return out
}

// jsonCompact is json.dumps(..., sort_keys=True, separators=(",", ":"), ensure_ascii=False).
func jsonCompact(v any) string {
	var b strings.Builder
	var write func(any)
	write = func(v any) {
		switch t := v.(type) {
		case Obj:
			sorted := append(Obj(nil), t...)
			sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].Key < sorted[j].Key })
			b.WriteByte('{')
			for i, f := range sorted {
				if i > 0 {
					b.WriteByte(',')
				}
				writeRawString(&b, f.Key)
				b.WriteByte(':')
				write(f.Value)
			}
			b.WriteByte('}')
		case []any:
			b.WriteByte('[')
			for i, x := range t {
				if i > 0 {
					b.WriteByte(',')
				}
				write(x)
			}
			b.WriteByte(']')
		case string:
			writeRawString(&b, t)
		default:
			b.WriteString(dumps(v))
		}
	}
	write(v)
	return b.String()
}

// writeRawString is a JSON string with ensure_ascii=False: only quote, backslash and controls escape.
func writeRawString(b *strings.Builder, s string) {
	b.WriteByte('"')
	for _, r := range s {
		switch {
		case r == '"' || r == '\\':
			b.WriteByte('\\')
			b.WriteRune(r)
		case r == '\n':
			b.WriteString(`\n`)
		case r == '\r':
			b.WriteString(`\r`)
		case r == '\t':
			b.WriteString(`\t`)
		case r == '\b':
			b.WriteString(`\b`)
		case r == '\f':
			b.WriteString(`\f`)
		case r < 0x20:
			fmt.Fprintf(b, `\u%04x`, r)
		default:
			b.WriteRune(r)
		}
	}
	b.WriteByte('"')
}

// SetDigest is criteria.set_digest: canonical JSON, so no delimiter can collide two sets.
func SetDigest(cs []Criterion) string {
	sorted := slices.Clone(cs)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].ID < sorted[j].ID })
	sum := sha256.Sum256([]byte(jsonCompact(criteriaObjs(sorted))))
	return hex.EncodeToString(sum[:])
}

// NormaliseCriteria is normalise_criteria: the one shape a set is stored in.
func NormaliseCriteria(entries []any) ([]Criterion, error) {
	var out []Criterion
	seen := map[string]bool{}
	for _, entry := range entries {
		o, ok := entry.(Obj)
		if !ok {
			return nil, refuse(CriteriaUnregistered, "each criterion is an object with an id and a title")
		}
		idv, _ := get(o, "id")
		titlev, _ := get(o, "title")
		id := strings.TrimSpace(pyStrOrEmpty(idv))
		title := strings.TrimSpace(pyStrOrEmpty(titlev))
		if id == "" || title == "" {
			return nil, refuse(CriteriaUnregistered, "each criterion needs a non-empty id and title")
		}
		if seen[id] {
			return nil, refuse(CriteriaUnregistered, "duplicate criterion id %s", store.PyRepr(id))
		}
		seen[id] = true
		required := true
		if r, present := get(o, "required"); present {
			required = truthy(r)
		}
		out = append(out, Criterion{id, title, required})
	}
	if len(out) == 0 {
		return nil, refuse(CriteriaUnregistered, "a criteria set needs at least one criterion")
	}
	return out, nil
}

// pyStrOrEmpty is str(value or "").
func pyStrOrEmpty(v any) string {
	if !truthy(v) {
		return ""
	}
	return pyStr(v)
}

// NormaliseFindings is normalise_findings: criteria and findings merged by id, dispositions in
// the frozen enum, at most one restoration carrier and never an empty one.
func NormaliseFindings(sources ...[]any) ([]any, error) {
	var merged []Obj
	declared := map[string]bool{}
	declaredSet := map[string]bool{}
	for _, source := range sources {
		for _, item := range source {
			o, ok := item.(Obj)
			if !ok {
				return nil, refuse(DispositionConflict, "each finding is an object")
			}
			idv, _ := get(o, "id")
			id := strings.TrimSpace(pyStrOrEmpty(idv))
			if id == "" {
				return nil, refuse(DispositionConflict, "each finding names a criterion id")
			}
			disposition := "verified"
			if v, _ := get(o, "verdict"); truthy(v) {
				disposition = pyStr(v)
				if _, isText := v.(string); !isText || !slices.Contains(dispositions, disposition) {
					return nil, refuse(DispositionConflict, "%s is not one of ('verified', 'needs_changes', 'unverified'); the contract's criteria enum is frozen and a finding outside it cannot be recorded", pyReprValue(v))
				}
			}
			entry := Obj{{Key: "id", Value: id}, {Key: "verdict", Value: disposition}}
			notev, _ := get(o, "note")
			if note := strings.TrimSpace(pyStrOrEmpty(notev)); note != "" {
				entry = append(entry, F{Key: "note", Value: note})
			}
			if flag, present := get(o, "restoration"); present && flag != nil {
				b, isBool := flag.(bool)
				if !isBool {
					return nil, refuse(DispositionConflict, "a finding declares its restoration block with true or false, not %s", pyTypeName(flag))
				}
				if declaredSet[id] && declared[id] != b {
					return nil, refuse(DispositionConflict, "%s both declares and disclaims the restoration block; one correction carries one block and says so once", store.PyRepr(id))
				}
				declared[id], declaredSet[id] = b, true
			}
			if declared[id] {
				entry = append(entry, F{Key: "restoration", Value: true})
			}
			merged = slices.DeleteFunc(merged, func(e Obj) bool { return str(e, "id") == id })
			merged = append(merged, entry)
		}
	}
	var carriers []string
	for _, e := range merged {
		if v, _ := get(e, "restoration"); v == true {
			carriers = append(carriers, str(e, "id"))
		}
	}
	if len(carriers) > 1 {
		return nil, refuse(DispositionConflict, "%s each declare the restoration block. One correction carries one block, and two candidates is a block nobody can locate", reprList(carriers))
	}
	for _, e := range merged {
		if v, _ := get(e, "restoration"); v == true {
			if _, hasNote := get(e, "note"); !hasNote {
				return nil, refuse(DispositionConflict, "%s declares the restoration block and carries no note. The block is the note; a declaration without one names a carrier with nothing in it", store.PyRepr(str(e, "id")))
			}
		}
	}
	out := make([]any, len(merged))
	for i, e := range merged {
		out[i] = e
	}
	return out, nil
}

// Criteria is criteria.CriteriaService.
type Criteria struct {
	Store *store.Store
	Clock Clock
}

func (c *Criteria) registration(rid string, cs []Criterion, digest string, source any) Obj {
	return Obj{{Key: "relationshipId", Value: rid}, {Key: "setDigest", Value: digest}, {Key: "sourceRef", Value: source}, {Key: "criteria", Value: criteriaObjs(cs)}, {Key: "mode", Value: Managed}}
}

// Register is register: an intentional replacement.
func (c *Criteria) Register(ctx context.Context, rid string, entries []any, source any) (Obj, error) {
	cs, err := NormaliseCriteria(entries)
	if err != nil {
		return nil, err
	}
	digest, now := SetDigest(cs), c.Clock.ISO()
	err = c.Store.Transaction(ctx, func(ctx context.Context, _ *sql.Conn) error { return c.replace(ctx, rid, cs, digest, source, now) })
	if err != nil {
		return nil, err
	}
	return c.registration(rid, cs, digest, source), nil
}

// EnsureRegistered is ensure_registered: an exact replay is a no-op; a different set refuses.
func (c *Criteria) EnsureRegistered(ctx context.Context, rid string, entries []any, source any) (Obj, error) {
	cs, err := NormaliseCriteria(entries)
	if err != nil {
		return nil, err
	}
	digest, now := SetDigest(cs), c.Clock.ISO()
	err = c.Store.Transaction(ctx, func(ctx context.Context, _ *sql.Conn) error {
		existing, err := c.lockedSet(ctx, rid)
		if err != nil {
			return err
		}
		if existing == nil {
			return c.replace(ctx, rid, cs, digest, source, now)
		}
		sorted := slices.Clone(cs)
		sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].ID < sorted[j].ID })
		if !slices.Equal(existing.criteria, sorted) || existing.digest != digest || existing.source != source || existing.mode != Managed {
			return refuse(CriteriaSetChanged, "criteria for %s are already registered as %s; ensure_registered does not replace them", store.PyRepr(rid), existing.digest)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	return c.registration(rid, cs, digest, source), nil
}

type storedSet struct {
	criteria []Criterion
	digest   string
	source   any
	mode     string
}

func (c *Criteria) lockedSet(ctx context.Context, rid string) (*storedSet, error) {
	rows, err := all(ctx, c.Store, "SELECT criterion_id, title, required, source_ref, set_digest, recorded_at FROM canonical_criteria WHERE relationship_id = ? ORDER BY criterion_id", rid)
	if err != nil {
		return nil, err
	}
	mode, err := one(ctx, c.Store, "SELECT mode FROM verification_mode WHERE relationship_id = ?", rid)
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 && mode == nil {
		return nil, nil
	}
	var cs []Criterion
	sources, digests := map[any]bool{}, map[string]bool{}
	for _, row := range rows {
		if r := row.I("required"); r != 0 && r != 1 {
			return nil, refuse(CriteriaSetChanged, "stored required flag for %s is not 0 or 1", store.PyRepr(row.S("criterion_id")))
		}
		cs = append(cs, Criterion{row.S("criterion_id"), row.S("title"), row.I("required") == 1})
		sources[row.Opt("source_ref")] = true
		digests[row.S("set_digest")] = true
	}
	if len(cs) == 0 || mode == nil || mode.S("mode") != Managed || len(sources) != 1 || len(digests) != 1 || !digests[SetDigest(cs)] {
		return nil, refuse(CriteriaSetChanged, "criteria for %s are stored in a form ensure_registered will not replace or repair", store.PyRepr(rid))
	}
	return &storedSet{cs, rows[0].S("set_digest"), rows[0].Opt("source_ref"), mode.S("mode")}, nil
}

func (c *Criteria) replace(ctx context.Context, rid string, cs []Criterion, digest string, source any, now string) error {
	if _, err := execSQL(ctx, c.Store, "DELETE FROM canonical_criteria WHERE relationship_id = ?", rid); err != nil {
		return err
	}
	for _, e := range cs {
		required := int64(0)
		if e.Required {
			required = 1
		}
		if _, err := execSQL(ctx, c.Store, "INSERT INTO canonical_criteria (relationship_id, criterion_id, title, required, source_ref, set_digest, recorded_at) VALUES (?,?,?,?,?,?,?)", rid, e.ID, e.Title, required, source, digest, now); err != nil {
			return err
		}
	}
	if err := c.writeMode(ctx, rid, Managed, now); err != nil {
		return err
	}
	return journal(ctx, c.Store, "criteria_registered", rid, Obj{{Key: "setDigest", Value: digest}, {Key: "count", Value: int64(len(cs))}}, now)
}

func (c *Criteria) writeMode(ctx context.Context, rid, mode, now string) error {
	_, err := execSQL(ctx, c.Store, "INSERT INTO verification_mode (relationship_id, mode, recorded_at) VALUES (?,?,?) ON CONFLICT(relationship_id) DO UPDATE SET mode = excluded.mode, recorded_at = excluded.recorded_at", rid, mode, now)
	return err
}

// SetMode is set_mode.
func (c *Criteria) SetMode(ctx context.Context, rid, mode string) (Obj, error) {
	if mode != Managed && mode != Legacy {
		return nil, refuse(DispositionConflict, "unknown verification mode %s", store.PyRepr(mode))
	}
	err := c.Store.Transaction(ctx, func(ctx context.Context, _ *sql.Conn) error { return c.writeMode(ctx, rid, mode, c.Clock.ISO()) })
	return Obj{{Key: "relationshipId", Value: rid}, {Key: "mode", Value: mode}}, err
}

// Get is get: the registered set, or nil.
func (c *Criteria) Get(ctx context.Context, rid string) (Obj, error) {
	rows, err := all(ctx, c.Store, "SELECT * FROM canonical_criteria WHERE relationship_id = ? ORDER BY criterion_id", rid)
	if err != nil || len(rows) == 0 {
		return nil, err
	}
	var cs []any
	for _, r := range rows {
		cs = append(cs, Criterion{r.S("criterion_id"), r.S("title"), r.I("required") != 0}.obj())
	}
	return Obj{{Key: "relationshipId", Value: rid}, {Key: "setDigest", Value: rows[0].S("set_digest")}, {Key: "sourceRef", Value: rows[0].Opt("source_ref")}, {Key: "criteria", Value: cs}}, nil
}

// Mode is mode.
func (c *Criteria) Mode(ctx context.Context, rid string) (string, error) {
	row, err := one(ctx, c.Store, "SELECT mode FROM verification_mode WHERE relationship_id = ?", rid)
	if err != nil || row == nil {
		return Legacy, err
	}
	return row.S("mode"), nil
}

// BindReview is bind_review: the set's digest as it stands at review start.
func (c *Criteria) BindReview(ctx context.Context, rid, eventID string) error {
	current, err := c.Get(ctx, rid)
	if err != nil {
		return err
	}
	var digest any
	if current != nil {
		digest = str(current, "setDigest")
	}
	_, err = execSQL(ctx, c.Store, "INSERT OR IGNORE INTO claim_context (event_id, set_digest, bound_at) VALUES (?,?,?)", eventID, digest, c.Clock.ISO())
	return err
}

// BoundDigest is bound_digest.
func (c *Criteria) BoundDigest(ctx context.Context, eventID string) (any, error) {
	row, err := one(ctx, c.Store, "SELECT set_digest FROM claim_context WHERE event_id = ?", eventID)
	if err != nil || row == nil {
		return nil, err
	}
	return row.Opt("set_digest"), nil
}

// Coverage is coverage: the verdict's findings against the canonical set.
func (c *Criteria) Coverage(ctx context.Context, rid, eventID, verdict string, findings []any, reason any, expected any) (Obj, error) {
	registered, err := c.Get(ctx, rid)
	if err != nil {
		return nil, err
	}
	mode, err := c.Mode(ctx, rid)
	if err != nil {
		return nil, err
	}
	if findings == nil {
		findings = []any{}
	}
	if registered == nil {
		if mode == Managed {
			return nil, refuse(CriteriaUnregistered, "%s is a managed assignment with no canonical criteria; a managed assignment cannot be completed against nothing", store.PyRepr(rid))
		}
		return Obj{{Key: "coverage", Value: LegacyUnregistered}, {Key: "setDigest", Value: nil}, {Key: "boundDigest", Value: nil}, {Key: "findings", Value: findings}}, nil
	}
	digest := str(registered, "setDigest")
	bound, err := c.BoundDigest(ctx, eventID)
	if err != nil {
		return nil, err
	}
	if mode == Managed && bound == nil && expected == nil && (verdict == "verified" || verdict == "needs_changes") {
		return nil, refuse(ReviewNotBound, "this managed review is not bound to a criteria set: claim the event first, or pass the reviewed digest explicitly. Current set is %s", digest)
	}
	if bound != nil && bound != digest {
		return nil, refuse(CriteriaSetChanged, "the criteria set changed after this review was claimed: bound %s, current %s. Findings made against the previous wording cannot certify the current one; claim the review again", pyStr(bound), digest)
	}
	if expected != nil && expected != digest {
		return nil, refuse(CriteriaSetChanged, "expected criteria set %s, but the current set is %s", pyStr(expected), digest)
	}
	list, _ := get(registered, "criteria")
	known, required := map[string]bool{}, []string{}
	for _, x := range list.([]any) {
		o := x.(Obj)
		known[str(o, "id")] = true
		if v, _ := get(o, "required"); v == true {
			required = append(required, str(o, "id"))
		}
	}
	byID := map[string]Obj{}
	for _, f := range findings {
		o := f.(Obj)
		if !known[str(o, "id")] {
			return nil, refuse(UnknownCriterion, "%s is not in this assignment's canonical criteria", store.PyRepr(str(o, "id")))
		}
		byID[str(o, "id")] = o
	}
	hasNote := func(o Obj) bool { v, _ := get(o, "note"); return truthy(v) }
	switch verdict {
	case "verified":
		var missing []string
		for _, id := range required {
			if o, ok := byID[id]; !ok || str(o, "verdict") != "verified" {
				missing = append(missing, id)
			}
		}
		sort.Strings(missing)
		if len(missing) > 0 {
			return nil, refuse(CriteriaNotCovered, "these required criteria are not recorded as verified: %s", reprList(missing))
		}
	case "needs_changes":
		ok := false
		for _, o := range byID {
			if str(o, "verdict") == "needs_changes" && hasNote(o) {
				ok = true
			}
		}
		if !ok {
			return nil, refuse(FindingsRequired, "a needs_changes verdict needs at least one criterion marked needs_changes with a note; a correction with no findings is one nobody can act on")
		}
	case "unverified":
		ok := false
		for _, o := range byID {
			if str(o, "verdict") == "unverified" && hasNote(o) {
				ok = true
			}
		}
		if !ok {
			return nil, refuse(FindingsRequired, "an unverified verdict names at least one criterion it could not verify, with the reason")
		}
	case "aborted":
		if strings.TrimSpace(pyStrOrEmpty(reason)) == "" {
			return nil, refuse(FindingsRequired, "an aborted verdict states why")
		}
	}
	return Obj{{Key: "coverage", Value: Covered}, {Key: "setDigest", Value: digest}, {Key: "boundDigest", Value: bound}, {Key: "findings", Value: findings}}, nil
}
